design/src/rfcs/rfc0025_better_constraint_violations.md

```diff
@@ -0,0 +1,511 @@
+deserializers enforce constraint traits in a two-step phase: first, the
+_entirety_ of the unconstrained value is deserialized, _then_ constraint traits
+are enforced by feeding the entire value to the `TryFrom` constructor.
+
+We will introduce a `ConstraintViolations` type (note the plural) that
+represents a collection of constraint violations that can occur _within user
+application code_. Roughly:
+
+```rust
+pub ConstraintViolations<T>(pub(crate) Vec<T>);
+
+impl<T> IntoIterator<Item = T> for ConstraintViolations<T> { ... }
+
+impl std::convert::TryFrom<std::string::String> for LengthPatternString {
+    type Error = ConstraintViolations<crate::model::length_pattern_string::ConstraintViolation>;
+
+    fn try_from(value: std::string::String) -> Result<Self, Self::Error> {
+        // Check constraints and collect violations.
+        ...
+    }
+}
+```
+
+- The main reason for wrapping a vector in `ConstraintViolations` as opposed to
+  directly returning the vector is forwards-compatibility: we may want to
+  expand `ConstraintViolations` with conveniences.
+- If the constrained type can only ever yield a single violation, we will
+  dispense with `ConstraintViolations` and keep directly returning the
+  `crate::model::shape_name::ConstraintViolation` type.
+
+We will analogously introduce a `ConstraintViolationExceptions` type that
+represents a collection of constraint violations that can occur _within the
+framework's request deserialization code_. This type will be `pub(crate)` and
+will be the one the framework will map to Smithy's `ValidationException` that
+eventually gets serialized into the response.
+
+### Unresolved questions
+
+#### Collecting constraint violations may constitute a DOS attack vector
+
+This is a problem that _already_ exists as of writing, but that collecting
+constraint violations highlights, so it is a good opportunity, from a
+pedagogical perspective, to explain it here. Consider the following model:
+
+```smithy
+@length(max: 3)
+list ListOfPatternStrings {
+    member: PatternString
+}
+
+@pattern("expensive regex to evaluate")
+string PatternString
+```
+
+Our implementation currently enforces constraints _from the leaf to the root_:
+when enforcing the `@length` constraint, the `TryFrom` constructor the server
+framework uses gets a `Vec<String>` and _first_ checks the members adhere to
+the `@pattern` trait, and only _after_ is the `@length` trait checked. This
+means that if a client sends a request with `n >>> 3` list members, the
+expensive check runs `n` times, when a constant-time check inspecting the
+length of the input vector would have sufficed to reject the request.
+Additionally, we may want to avoid serializing `n` `ValidationExceptionField`s
+due to performance concerns.
+
+- A possibility to circumvent this is making the `@length` validator special,
+  having it bound the other validators via effectively permuting the order of
+  the checks and thus short-circuiting.
```

Comment:
I am not sure I understand why this applies only to `@length`. Shouldn't this issue apply to basically any of the constraints?

Reply:
I don't see why it should apply to any of the other constraints. As the document mentions, I think that `@uniqueItems` on list shapes and `@length` on list shapes are the only ones that should short-circuit. I think there is no general rule to determine whether a constraint should short-circuit or not; it has to be determined on a case by case basis.